src/pinyin.cpp

```diff
@@ -0,0 +1,493 @@
+    {U"zhua", U"抓摣撾檛爪爫膼髽𠆣𡎬𣑃𥬲𥮣"},
+    {U"zhuai", U"尵拽拽睉跩顡𢶀"},
+    {U"zhuan", U"㛯㼷䉵䏙䏝䡱䧘专傳僎叀啭囀堟塼嫥孨専專恮撰沌灷瑑瑼甎砖磗磚竱篆篹篿籑縳耑腞膞膞蒃蟤襈諯譔賺赚転轉轉转鄟顓颛饌馔鱄𠊩𠨎𡇰𡢀𡤛𡩚𡭇𡰞𢂘𢐎𢞬𣂵𣕏𣚢𤂤𤩄𤪪𤮳𥛥𥫛𦁆𦄯𦓝𦝏𦠆𦧸𧂍𧶲𧸖𨷱𩔊𩧜𩳏𩻝"},
+    {U"zhuang", U"䚒僮壮壯壵奘妆妝娤幢庄庒戇撞桩梉樁湷漴焋状狀獞粧糚荘莊装裝𠌴𢙳𢤤𢦅𣞝𣴣𣶍𣻛𤕯𤘲𤶜𦀜𦚏𩅃𩮱𩯲𪁈𪄻𪅖𪉉"},
+    {U"zhui", U"㗓㚝㩾㮅㾽䄌䢡䧳䨨䳡䶆倕坠墜娷惴桘椎沝甀硾磓礈笍綴縋缀缒腏膇萑諈譵贅赘轛追醀錐錗錣鑆锥隹餟騅骓鵻𠄌𡑻𡯑𢊅𣝸𣦬𣨫𤺅𥟒𦥻𧿲𨒉𨧨𨪗𨺵𨾻𩛵𩜀𩪀𩬳𪋇"},
+    {U"zhun", U"㡒䐃准凖圫埻宒屯忳準稕稕窀綧綧肫衠訰訰諄谆迍𥇜𥚠"},
+    {U"zhuo", U"㑁㒂㓸㣿㧳㧻㪬㭬㹿㺟䂐䃗䅵䆯䐁䓬䕴䖦䞵䟾䦃䪼䫎䮓䮕䵠䵵䶂丵倬剢劅卓叕啄噣圴墌妰娺彴拙捉捔撯擆擢斀斫斮斱斲斵晫桌梲棳棹椓槕櫡汋浊浞涿濁濯灂灼炪烵焯焯犳狵琢琸着矠硺禚穛穱窡窧篧籗籱罬聉茁著蓔蠗蠿諁諑謶诼趠酌鉵鋜鐯鐲镯鷟𠡑𠭴𠿡𡷿𢁁𢢗𢧈𢳇𢺡𢽚𣃈𣃑𣄻𤃮𤉐𤏸𤓦𥇍𥋮𥐊𥗁𥞺𥢔𥮥𥯩𥷘𥷮𥼚𦜰𦰹𦳡𧂒𧃔𧘑𧞐𧢼𧨳𧰵𧱒𧱰𨁿𨑽𨖮𨡸𨢬𨧧𨮿𨺝𩆸𩋁𩑂𩩔𩲃𩷹"},
+    {U"zi", U"㜽㞨㠿㧗㧘㬐㰣㰷㱴㺭㽧㾅㿳䁒䅆䅔䆅䎩䐉䔂䖪䘣䣎䦻䰵乲仔仔倳兹剚吇吱呰呲咨啙嗞姊姕姿子子孖字孜孳孶崰嵫恣扻杍栥梓椔榟橴沝洓淄渍湽滋滓漬澬牸玆璾眥眦矷禌秄秄秭秶稵笫籽粢紎紫緇緕缁耔胏胔胾自芓芓芓茈茊茡茲荢菑葘蓻薋虸觜訾訾訿諮谘貲資赀资趑趦輜輺辎鄑釨鈭錙鍿鎡锱镃頾頿髭鯔鰦鲻鶅鶿鼒齍齜龇𠀢𠂔𠡸𡉗𡗈𡙛𡥎𡪒𡸟𡸪𢆶𢱆𢼱𣄮𣓊𣖨𣚀𣚁𣣊𣣌𣥨𣯃𣳩𣸆𤲸𥀖𥕁𥚉𥞎𥫞𥬳𥲕𥻍𥼩𥼻𥿩𦍺𦎸𦖺𦞤𦣹𦺱𧀗𧂐𧆰𧕓𧙁𧛏𧣤𧥕𧧕𧨴𧹌𧿞𨀥𨍢𨚖𨝳𨧫𨩲𨹀𩄚𩐍𩜊𪅵𪑿𪕊𪗉𪗋𪗐"},
+    {U"zong", U"㙡㚇㢔㣭㨑㯶㷓㹅䁓䁓䈦䍟䑸䖲䗥䙕䝋䡯䢨䰌倊倧偬傯堫宗嵏嵕嵸從总惣惾愡捴揔搃摠昮朡枞棕椶潈潨熜熧猔猣疭瘲碂磫稯稯粽糉糭綜緃総緵緵縂縦縱總繌纵综翪腙艐葼蓗蝬豵踨踪蹤錝鏓鑁騌騣骔鬃鬉鬷鯮鯼𠏭𠕌𠡻𠵻𡕰𡞧𣀒𣯨𤡆𤡆𥍺𥓻𥚾𥠡𦖸𦡙𧺣𨌰𨍈𨍉𨎢𨺡𩋯𩤗𩦲𩭤𩭤𩮀𩮀𩰽𪖁"},
+    {U"zou", U"㔌㔿㵵㻓䅳䠫奏媰掫揍棷棸楱箃緅菆諏诹走赱邹郰鄒鄹陬騶驟驺鯐鯫鲰黀齱齺𠂑𣙻𣠏𥋜𥶈𧌗𨂡𨃘𨑿𨜗𨝮𨽁𩼦𪃆"},
+    {U"zu", U"㞺㧻㩆㰵㲞㵀䃚䅸䔃䖕䙘䚝䯿䱣伹俎傶卆卒哫崒崪捽族爼珇祖租稡箤組綷组菹葅蒩詛謯诅足踿鉃鉐錊鎐鎺鏃镞阻靻𠻏𡻬𢅪𢉺𢫵𢳇𢳈𣇙𣢰𣤶𣨛𤓵𤬧𤱌𤽱𥛜𥞯𥣆𥼀𥼪𦑋𦵬𧇈𧇿𧎲𧐈𧑙𧗎𧞰𧺒𨂀𨃭𨄕𨧰𨨳𨩰𩐡𩥿𩩠𩲲𩺯𪋍𪘧𪙳"},
+    {U"zuan", U"㸇䂎䌣䡽䤸䰖劗揝攥欑籫繤纂纉纘缵賺躜躦鑚鑽鑽钻𡉺𣀶𣪁𦆈𦙉𨇃𨉖𨰭𩎑𩛻𩯳"},
+    {U"zui", U"㝡㠑㭰㭰㰎䓱䘒䘹䮔䯿厜嗺嗺嘴噿嫢嶉嶊嶵晬最栬槜槯樶檇檌欈濢璻祽稡絊纗罪蕞蟕觜辠酔酨酻醉鋷錊𠲋𠾋𠿘𡙭𡡔𡽁𡽛𢈡𢊛𣖱𣩑𥍋𥳣𥳵𦆈𦈬𦏳𦙈𦸺𧎹𧻝𨢅𨻵𨿇𩚻𩣷𩲨𪋌𪓌"},
+    {U"zun", U"䔿僔噂墫壿尊嶟拵捘撙樽繜罇袸譐遵銌鐏鱒鳟鶎鷷𠟃𠱜𤮐𥊭𥞘𥢎𦢐𦨆𦪚𨱔𩯄"},
+    {U"zuo", U"㑅㘀㘴㘸㛗㝾㤰㩇㭮㵶㸲䋏䎰䔘䝫䞢䞰䟄䟭䟶䦈佐作侳做咗唑唨嘬坐岝岞左座怍捽昨柞椊祚秨稓穝筰糳繓胙苲莋葃葄蓙袏諎酢鈼鑿阼飵𠂇𠱯𠹠𡪳𡯨𡹥𢀡𢂃𢤎𣠹𣩈𣹧𤿀𥅁𥙀𥥏𥹁𥽿𦁎𦈛𦥬𦦹𧃘𧨀𧮙𧲭𨀨𨐳𨝨𨞒𩛠𪎇𪎲"},
+    {U",", U"，"},
+    {U".", U"。"},
+    {U"?", U"？"},
+    {U"!", U"！"},
+    {U"\"", U"“”"}
+};
+
+void append_one_char( char32_t character, std::vector<std::u32string> &dest )
+{
+    std::vector<std::u32string> result; //result for this character
+    for( auto const &entry : pinyin_data ) {
+        if( entry.second.find( character ) != std::u32string::npos ) {
+            result.push_back( entry.first );
+        }
+    }
+
+
+    // no match, this character is unknown, therefore we just append it to the destination
+    if( result.size() == 0 ) {
+        for( std::u32string &existing_dest_str : dest ) {
+            existing_dest_str += character;
+        }
+        return;
+    }
+
+    // there is nothing in destination yet, just push the result into it
+    if( dest.size() == 0 ) {
+        for( std::u32string &result_entry : result ) {
+            dest.push_back( result_entry );
+        }
+        return;
+    }
+
+    // else, there is already 1 or more entry in destination
+    // therefore append all possible combinations
+    std::vector<std::u32string> new_dest;
+
+    for( std::u32string &existing_dest_str : dest ) {
+        for( std::u32string &result_entry : result ) {
+            new_dest.push_back( existing_dest_str + result_entry );
+        }
+    }
+    dest = new_dest;

```

Comment:
Added a cache that is lazy-loaded when a string is processed once, so subsequent search will be faster. The cache then get reset when the world is unloading in case the user decide to load a different world with different mods.

It will still be slow on first search, but assuming a user does not search once per play session, it should overall bring some improvements.

Reply:
Well, caching does improve the speed but is going to exacerbate the memory problem because now the memory consumption not only increases exponentially with the number of multi-pronunciation characters in each string but also linearly with the number of searched strings... What I meant by lazy evaluation is that this list of possible combinations doesn't have to be computed all at once; you could calculate one at a time and compute the next only if the previous one does not match the query string. I don't know if that is needed though, maybe there are not enough edge case strings to trigger the problems I described.

Reply:
Actually that is a good idea, I will try to do that, thanks for the advice!